core/src/main/scala/org/apache/spark/scheduler/Stage.scala

```diff
@@ -69,7 +69,21 @@ private[spark] abstract class Stage(
 
   /** Pointer to the latest [StageInfo] object, set by DAGScheduler. */
   var latestInfo: StageInfo = StageInfo.fromStage(this)
+  
+  // Spark is resilient to executors dying by retrying stages on FetchFailures. This counter ensures
+  // that we don't retry stages indefinitely by aborting the stage if it fails too many times. 
+  private var failCount = 0
+  private[scheduler] def fail() : Unit = { failCount += 1 }
+  private[scheduler] def shouldAbort(): Boolean = { failCount >= Stage.maxStageFailures }
+  private[scheduler] def clearFailures() : Unit = { failCount = 0 }
 
+  /**
+   * Check whether we should abort the failedStage due to multiple failures.
+   * This method updates the running count of failures for a particular stage and returns 
+   * true if the number of failures exceeds the allowable number of failures.
+   */
+  private[scheduler] def failAndShouldAbort(): Boolean = { fail(); shouldAbort() }
```

Comment:
errr ... sorry @ilganeli , I meant something different by inlining them.  I meant to get rid of `fail()` and `shouldAbort()` as separate methods, and have this method be just:

```
private[scheduler] def failAndShouldAbort(): Boolean = {
  failCount += 1
  failCount >= Stage.maxStageFailures
}
```
since they are so simple, just for readability.